### DISCUSSION TITLE
CarPlay / Android Auto support

### DISCUSSION THREAD

**Discussion by angeloreale** [expo/expo#24354] [Proposals and Ideas]
**Context**
I have a radio app running in Expo ecosystem, and I don't want to leave Expo ecosystem to have my radio app playable in vehicles.

I'm starting to gather requirements to write an RFC / implementation for adding support for CarPlay / Android Auto interfaces into Expo.

**Progress**
[#-------------------] 5%

So far, I could find that there is at least a crucial feature missing in Expo which seems to be a way to leverage custom AppScenes / UIApplicationDelegate interfaces.

2023-09-10: I'm trying to understand whether Expo makes any usage of Scenes and whether it'd require tweaking of enablement. [Ref 1]

Also, as a natural follow up, at least for my use case, would be to have the library React Native Track Player being able to reproduce on a/any car sound system, but I do see that there is some current progress in there too. [Ref 2] 

Ref 1: https://github.com/birkir/react-native-carplay/issues/101
Ref 2: https://github.com/doublesymmetry/react-native-track-player/pull/2094

**Comment by tsapeta** [comment]
Hey @angeloreale 👋 
We currently don't use scenes anywhere, but we're open for an RFC and adding support for them. Also @knowbody might be interested in helping on this!

**Reply by forki** [reply]
@EvanBacon I remember you wrote on twitter that react-native-carplay should work!?

**Reply by betocolon23** [reply]
Hi, any updates on expo and carplay/android auto integration? I have a fully developed app on expo and want to do some things with car play :(

**Reply by philipkaare** [reply]
Any updates?

**Reply by simplenotezy** [reply]
@betocolon23 is it not possible for your to use https://github.com/birkir/react-native-carplay while still using expo?

**Comment by vanGalilea** [comment]
The need for CarPlay / Android Auto support grows with the day, can we bump the prio of it by Expo @EvanBacon?

**Comment by Jojocaster** [comment]
CarPlay would be  a lovely addition. I'm also using `react-native-track-player` and we could really use CarPlay / AndroidAuto. As others mentioned, this seems to be restricted by expo using Scenes. It would be great to know whether this is something that you plan on supporting or not, I would like to avoid going back to bare RN if possible. 

Really appreciate the work you're doing, thanks to the team!

**Comment by krsche** [comment]
Would also love to have support for Android Auto / CarPlay. With demand for this increasing every day, it would be something that hugely adds to Expo's attractiveness as a framework.

**Comment by nixolas1** [comment]
Build once, run everywhere :)))))

**Comment by tmetzger** [comment]
I'll add my voice to this. Would love to add carplay/android auto support to my expo app.

**Comment by korysericati** [comment]
CarPlay is the only obstacle preventing me from pulling all of my clients' projects into Expo/EAS by default. It's easy enough to tack on Android Auto with a plugin but the lack of support for scenes makes CarPlay a nightmare. Has there been any movement on this?

**Comment by ryderhorne** [comment]
Hello! Are there any updates on this? The company I’m working at a company that wants to add support for their mobile application in Apple Carplay and Android Auto, but we also plan to use Expo, so having support for it within Expo would be great!

**Comment by ser-emejia** [comment]
Please Expo team, make it possible 🙏

**Comment by valentin-debris** [comment]
Hi,
I'm starting my first app with Expo (with `npx create-expo-app@latest`) but I'm generating the android/ios folder because I need to add custom stuff (with `npx expo prebuild`). Does this means I will be able to use `react-native-carplay` or should I have to create the app as Bare app (with something like `npx @react-native-community/cli@X.XX.X init`) ?
Regards

**Comment by nathancovey** [comment]
Is there a workaround currently to get this working in CarPlay with Expo?

**Comment by The4star** [comment]
Getting carplay working for expo would be amazing for [Kuratour](https://kuratour.com), how can we get this going?

**Comment by Crease29** [comment]
I'd also absolutely love to see this coming 🤩

**Comment by g4rb4g3** [comment]
Hello everyone! I have forked react-native-carplay quite a while ago and added Dashboard and Cluster scenes. We're running this on a production navigation app using expo SDK 52 currently. Android Auto is in the making, check the latest commits for updates.
https://github.com/g4rb4g3/react-native-carplay
Once AA is done I'll update all the docs to give details on how to get the scenes working on Expo.

**Reply by g4rb4g3** [reply]
I also have a PR in place on rnmapbox/maps to get the gestures working nicely with my fork of react-native-carplay. There are screen recordings included showing CarPlay and Android Auto using my fork and the rnmapbox PR.
https://github.com/rnmapbox/maps/pull/3854

**Reply by forki** [reply]
That sounds exciting. I also have couple of fixes for AndroidAuto. I'll try to contribute that to your fork.

**Reply by LukasB-DEV** [reply]
@g4rb4g3 nice work. Thanks a lot. I am happy to see the further progress. If you need any help at testing or so feel free to reach out!

**Reply by nathancovey** [reply]
Great work!!

**Reply by The4star** [reply]
This is Fantastic news, I definitely owe you some beers! Keen as a bean for when it's ready to go.

**Reply by valentin-debris** [reply]
Great news ! Is your fork will works with project without Expo ?

**Reply by g4rb4g3** [reply]
@valentin-debris Yes I think so, keep in mind that you still have to find a way to make scenes work though.

**Reply by SteveWaterfall** [reply]
Hey - I also have CarPlay working with Expo builds (after a lot of pain), but I am struggling to get it working with a dev client build which would be great to be able to amend the CarPlay templates on the fly. I have tried various configurations in my eas.json to try and get it to work but no joy so far - Keep getting a black screen, or it just bypasses metro altogether and boots it up. If you have this working with a dev client I'd love to know how you did it! I have created a config plugin which does the necessary to the iOS directory.

**Reply by TerryMooreII** [reply]
Do either of you that have gotten Carplay to work with Expo have any detailed docs or can point me to a place that has doc on how to get this to work? Anything would helpful :) thanks and keep up the great work

**Reply by forki** [reply]
hey @g4rb4g3 any updates on this? Also I wonder if you could merge @birkir's master into your fork so that we don't have merge conflicts

**Reply by g4rb4g3** [reply]
@forki still working on AA, last commit was just yesterday so yes there are updates.

**Reply by forki** [reply]
unfortunately I can't comment on your fork with issues or send PRs. I was wondering if you already found the chance to update the docs a bit to match your changes and if you could try to merge the latest changes from @birkir into your fork to reduce merge conflicts. thanks

**Reply by g4rb4g3** [reply]
Tbh I have no idea why you are having troubles on doing PRs and issues, I got 3 PRs merged on my repo already so it seems to be an issue on your side.

One more thing, I'm working on a migration from a native Android Auto implementation to this react-native based one. So from my point of view I have more important things ongoing then updating docs and avoiding merge conflicts.
I'll do that later on when everything is done and working on our app.
So keep calm and just wait for things to happen.

**Reply by valentin-debris** [reply]
Hi @g4rb4g3 , while you're working on the improvement of this module, is it possible for you to share (or a sample) your AppDelegate.swift with the modification for the carPlay scne ? 
Based on my research, and depending on which RN version we use, everyone propose a different solution for this file (and PhoneScene/CarScene swift file), and since then, I'm having trouble to make it works properly.
Thanks again for the revival of this feature :)

**Reply by LukasB-DEV** [reply]
@g4rb4g3 i wanted to test and also pr. Can you open Discussions or Issues for your repository? I currently have the issue that the file react_native_carplay/react_native_carplay-Swift.h from ios/RNCarPlay.m is not there

**Reply by nixolas1** [reply]
Looking forward to your progress on the fork @g4rb4g3!
One heads up, in Expo 53 / RN 79 New Architechture is on by default, which uses TurboModules and JSI. Seems like the MapTemplate and probably NavigationTemplate won't work with the old RTCBridge with New Arch.
See issue here https://github.com/birkir/react-native-carplay/issues/225#issuecomment-3048972847

**Reply by Steffi3rd** [reply]
Hello @g4rb4g3, issues tab is not available on your fork. We would like to post an issue but no way to do it.
Maybe you can configure to enable it?

Same error as @LukasB-DEV

The current error we have with Expo 52 (CNG) is: 
**Pods/react-native-carplay: 'react_native_carplay/react_native_carplay-Swift.h' file not found**
<img width="939" height="619" alt="Capture d’écran 2025-08-08 à 15 57 25" src="https://github.com/user-attachments/assets/e7c71668-c532-4f9e-971a-83d5c0369445" />
<img width="887" height="666" alt="Capture d’écran 2025-08-08 à 15 57 32" src="https://github.com/user-attachments/assets/8d950f71-dd01-4f4a-8c72-a8be8ea02846" />

**Reply by g4rb4g3** [reply]
I've enabled issues on the repo now but don't expect anything for now.
Currently I know that Expo 52 and Expo 53 using the old architecture are working.
I'll check new architecture next when I'm back from vacation. Also CNG might be worth looking into, right now it's quite cumbersome to get rncp working on iOS...

**Reply by Steffi3rd** [reply]
@g4rb4g3 Thank you!!! We saw now Issues tab. 

I founded the fix for this error by the way. I will post the solution on the issue. 

Have a nice holiday!!

**Reply by marinario-m** [reply]
@LukasB-DEV @g4rb4g3 @Steffi3rd Hello, I have the same issue, I really appreciate any help or guidance.

❌  Pods/react-native-carplay: 'react_native_carplay/react_native_carplay-Swift.h' file not found

**Reply by LukasB-DEV** [reply]
> @LukasB-DEV @g4rb4g3 @Steffi3rd Hello, I have the same issue, I really appreciate any help or guidance.
> 
> ❌ Pods/react-native-carplay: 'react_native_carplay/react_native_carplay-Swift.h' file not found

have you installed it via npm? `npm install @g4rb4g3/react-native-carplay`

**Reply by LukasB-DEV** [reply]
my solution was to add this to my podfile:

`#Carplay
    installer.pods_project.targets.each do |target|
      if target.name == 'react-native-carplay'
        target.build_configurations.each do |config|
          config.build_settings['SWIFT_VERSION'] = '5.0'
          config.build_settings['HEADER_SEARCH_PATHS'] ||= ['$(inherited)']
          config.build_settings['HEADER_SEARCH_PATHS'] << '${PODS_ROOT}/Headers/Public/React'
          config.build_settings['HEADER_SEARCH_PATHS'] << '${PODS_ROOT}/Headers/Public/React-Core'
          config.build_settings['HEADER_SEARCH_PATHS'] << '${PODS_ROOT}/Headers/Public/React-RCTFabric'
        end
      end
    end`
    
 within the `post_install do |installer|`

**Reply by marinario-m** [reply]
> my solution was to add this to my podfile:
> 
> `#Carplay
>     installer.pods_project.targets.each do |target|
>       if target.name == 'react-native-carplay'
>         target.build_configurations.each do |config|
>           config.build_settings['SWIFT_VERSION'] = '5.0'
>           config.build_settings['HEADER_SEARCH_PATHS'] ||= ['$(inherited)']
>           config.build_settings['HEADER_SEARCH_PATHS'] << '${PODS_ROOT}/Headers/Public/React'
>           config.build_settings['HEADER_SEARCH_PATHS'] << '${PODS_ROOT}/Headers/Public/React-Core'
>           config.build_settings['HEADER_SEARCH_PATHS'] << '${PODS_ROOT}/Headers/Public/React-RCTFabric'
>         end
>       end
>     end`
>     
>  within the `post_install do |installer|`

Thanks for your help, but this solution didn't work for me.

**Reply by mefjuu** [reply]
@g4rb4g3 The fork looks promising! Is it compatible with the latest RN? Do both Android Auto and Apple CarPlay work?

**Reply by g4rb4g3** [reply]
@mefjuu This is verified to work with Expo SDK 53 so react-native 0.79.6. Still old architecture only for now. Next step is to get new architecture support but tis will take some more time.

**Comment by nathancovey** [comment]
This would be so cool and make me love Expo even more @EvanBacon 🙏

**Comment by kkx64** [comment]
Just adding to all the comments, CarPlay via Expo would be a game changer for us

**Comment by tmetzger** [comment]
It makes me wonder, how hard is it to build a native module from scratch? I
mean these days, with the help of one of the many AI coding buddies...

On Sun, Jul 13, 2025 at 3:05 AM Kiril Krsteski ***@***.***>
wrote:

> Just adding to all the comments, CarPlay via Expo would be a game changer
> for us
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/expo/expo/discussions/24354#discussioncomment-13742939>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAELPGVSOKU3XJOQV6AIAML3IIVNJAVCNFSM6AAAAABQSTU3ZWVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTGNZUGI4TGOI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Tom Metzger ***@***.***)
"No problem can withstand the assault of sustained thinking."
  - Voltaire

**Comment by NicoM77** [comment]
Any updates? That would be a hugely important and really cool feature for Expo.

**Comment by tmetzger** [comment]
Ha ha... yeah I said I was wondering. I didn't say it was easy.

On Fri, Dec 5, 2025 at 11:48 AM amirghafouri ***@***.***>
wrote:

> @tmetzger <https://github.com/tmetzger> if it's so easy then why don't
> you do it with one of the many ai coding buddies?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/expo/expo/discussions/24354#discussioncomment-15176051>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAELPGWLTU2TRENI3Y3LDNT4AHOPPAVCNFSM6AAAAABQSTU3ZWVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTKMJXGYYDKMI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Tom Metzger ***@***.***)
"No problem can withstand the assault of sustained thinking."
  - Voltaire

**Comment by objectiveSee** [comment]
The `react-native-carplay` library is interesting, but the more important problem is supporting CarPlay at all, not rendering CarPlay with React syntax. If Expo supported iOS multi-scene apps (CarPlay scenes), developers could implement CarPlay entirely in native Swift while keeping the main app in Expo/RN, which would be simpler and more aligned with Apple’s APIs.

From here, what would be the best way to move this forward? Does this need a formal proposal or RFC, and if so, is there an example we should follow or an existing proposal we can build on? Happy to help organize or contribute if there’s a clear next step for getting this done.

**Reply by g4rb4g3** [reply]
Check out my new library, it works with Expo and requires only a minimal amount of native code to get it running.

https://github.com/Iternio-Planning-AB/react-native-auto-play

But I agree, react native in general has some quirks that make it harder to get everything up and running. And Expo adds some more... 😬

**Reply by objectiveSee** [reply]
@g4rb4g3 Thanks. Checking it out. How were you able to get your library working with Expo if there is no Expo support for scenes?

**Reply by g4rb4g3** [reply]
I just add the scenes and provide fixes for issues I found so far, check the known issues section for details. Btw it is not expo lacking support for scenes but a react-native thing in the end.

**Reply by objectiveSee** [reply]
Totally agree. Expo's in a tough spot dealing with both React Native core and the community. Props to the Expo team for managing that position.

I'll try out your library on a fresh project and see how it works with my setup.

**Reply by g4rb4g3** [reply]
Have fun! Keep in mind that its main focus is navigation so it definitely lacks some features not required for that topic. We are happy about PRs changing this... 😉

**Reply by bitcrumb** [reply]
@g4rb4g3 I think the biggest issue mainly lies with expo-dev-client being incompatble with scenes right? Do you mean you somehow managed to work around that?

**Reply by g4rb4g3** [reply]
No there is no workaround, this requires the bare workflow on expo to get it working.
It is mainly a react-native library that also works with Expo bare workflow.
For the Expo managed workflow it would be required to get the library into the Expo ecosystem. But for that several issues have to be sorted out first, scenes being the easiest one of them.
Check out the known issues on my lib to get an idea. Now imagine the unknown issues. 😉

**Reply by objectiveSee** [reply]
> @g4rb4g3 I think the biggest issue mainly lies with expo-dev-client being incompatble with scenes right? Do you mean you somehow managed to work around that?

@bitcrumb That's my understanding too. Not even the "dev client" incompatible, but the whole Expo system is built using AppDelegate not SceneDelegate. I'm actively trying to find a way to get around this while still using expo. Let me know if you find anything or if you want to share notes.

I tried https://github.com/Iternio-Planning-AB/react-native-auto-play but it didn't work for me with a fresh Expo project.

**Reply by g4rb4g3** [reply]
Just use the bare workflow on Expo, that way scenes can be used while still having the Expo framework.
There is currently no other way, react-native still lacks scenes support and expo builds on top of that.

**Reply by LukasB-DEV** [reply]
just implemented it with expo bare workflow and works like a charm! Thanks a lot for your work!

**Reply by janwiebe-jump** [reply]
it should be possible to create an expo config plugin to perform the native changes, without having to use expo bare workflow.

**Reply by g4rb4g3** [reply]
Not sure I wanna go down that road. The only way from my understanding would be to string manipulate AppDelegate which seems very brittle to me. ExpoAppDelegateSubscriber does not cover the needs for this lib.

It's just a few lines that have to be added but it depends a bit on your app, doing this manually will make sure you spot any error in time.

**Reply by bitcrumb** [reply]
You most certainly could (and I gobbled something together with the help of Claude Code), but using `expo-dev-client` will still result in a crash. It seems to assert all kinds of things related to the AppDelegate which clashes with the need of using Scenes for CarPlay.

**Reply by objectiveSee** [reply]
Unclear how difficult the Expo plugin would be. What are the actual changes needed to support it? It would be helpful if there was a definitive guide or list of changes that someone could follow. I'd be happy to work on the plugin if steps were listed somewhere.

**Reply by objectiveSee** [reply]
`expo-dev-client` is incompatible with this

**Reply by tonyberetini-burai** [reply]
This is why I am likely not going to use Expo for my new app since I need full support for CarPlay and Android auto. Vanilla React Native seams a better path forward unfortunately.

**Reply by g4rb4g3** [reply]
You can just use the expo bare workflow and everything will work out fine. We are on expo 54 using @iternio/react-native-auto-play and everything works fine. It needs some adjustments for some modules like expo-splash-screen though. But I guess that is more of a react-native issue in general as there is no scene support on iOS.

**Comment by DanielGe** [comment]
adding here that it would be wonderful if Expo supports CarPlay and Android Auto

**Comment by ofirk12** [comment]
So many requests about CarPlay means it is clearly needed.

Do we have any update on this? Is it in progress, planned for the future, or not on the roadmap yet?

We need to know so we can decide whether to wait or start looking for another solution.

**Reply by g4rb4g3** [reply]
react-native needs quite some patching and workarounds to get it working with CarPlay. On top of that some expo libs need patches too so I don't see this coming anytime soon.

You can get it working with the bare workflow though using https://github.com/Iternio-Planning-AB/react-native-auto-play
